TestAPIServiceCgroup fails occasionally

We've seen some occasional failures of this test on Jenkins:

```
14:25:39 === RUN   TestAPIServiceCgroup
14:25:46 --- FAIL: TestAPIServiceCgroup (6.90s)
14:25:46 	rkt_api_service_test.go:52: Running rkt api service
14:25:46 	rkt_tests.go:115: Running command: /home/admin/workspace/rkt-matrix/dummy/dummy1/label/debian-8-flavor-coreos/builds/build-rkt-coreos/build-rkt-1.8.0+git/tmp/functional/rkt --dir=/tmp/datadir-737106128 --local-config=/tmp/localdir-377601519 --system-config=/tmp/systemdir-876755586 --user-config=/tmp/userdir-171684857 --insecure-options=image run --interactive /home/admin/workspace/rkt-matrix/dummy/dummy1/label/debian-8-flavor-coreos/builds/build-rkt-coreos/build-rkt-1.8.0+git/tmp/functional/test-tmp/rkt-inspect-interactive.aci
14:25:46 	rkt_api_service_test.go:576: Pods are not in RUNNING state
14:25:46 	rkt_api_service_test.go:576: Pods are not in RUNNING state
14:25:46 	rkt_api_service_test.go:576: Pods are not in RUNNING state
14:25:46 	rkt_api_service_test.go:576: Pods are not in RUNNING state
14:25:46 	rkt_api_service_test.go:576: Pods are not in RUNNING state
14:25:46 	rkt_api_service_test.go:576: Pods are not in RUNNING state
14:25:46 	rkt_api_service_test.go:571: Pods are running
14:25:46 	rkt_api_service_test.go:156: API Pod info: id:"e363eae5-a67f-4e4c-ae8a-3f6b0f1aea0a" pid:10515 state:POD_STATE_RUNNING apps:<name:"rkt-inspect" image:<base_format:<type:IMAGE_TYPE_APPC version:"0.8.4" > id:"sha512-9ce35088a75eead90c15bd94736e0d3628c27a844398d0e3e9ff25e16242c878" > annotations:<Key:"appc.io/docker/repository" value:"whatever" > annotations:<Key:"coreos.com/rkt/stage1/run" value:"/ex/run" > annotations:<Key:"coreos.com/rkt/stage1/enter" value:"/ex/enter" > annotations:<Key:"coreos.com/rkt/stage1/gc" value:"/ex/gc" > > networks:<name:"default" ipv4:"172.16.28.2" > cgroup:"/user.slice/user-1000.slice/session-3.scope" created_at:1465568746015439946 started_at:1465568746143440523 
14:25:46 	rkt_api_service_test.go:159: Pod info: &{id:e363eae5-a67f-4e4c-ae8a-3f6b0f1aea0a pid:10515 state:running apps:map[rkt-inspect:0xc82027bea0] networks:map[default:0xc8203021a0] manifest:0xc820276c00 createdAt:1465568746015000000 startedAt:1465568746143000000}
14:25:46 	rkt_api_service_test.go:193: Expected the cgroup suffix to have "/machine-rkt\\x2de363eae5\\x2da67f\\x2d4e4c\\x2dae8a\\x2d3f6b0f1aea0a.scope", but saw "/user.slice/user-1000.slice/session-3.scope"
14:25:46 	rkt_api_service_test.go:156: API Pod info: id:"e363eae5-a67f-4e4c-ae8a-3f6b0f1aea0a" pid:10515 state:POD_STATE_RUNNING apps:<name:"rkt-inspect" image:<base_format:<type:IMAGE_TYPE_APPC version:"0.8.4" > id:"sha512-9ce35088a75eead90c15bd94736e0d3628c27a844398d0e3e9ff25e16242c878" name:"coreos.com/rkt-inspect" version:"1.8.0" > state:APP_STATE_RUNNING annotations:<Key:"appc.io/docker/repository" value:"whatever" > annotations:<Key:"coreos.com/rkt/stage1/run" value:"/ex/run" > annotations:<Key:"coreos.com/rkt/stage1/enter" value:"/ex/enter" > annotations:<Key:"coreos.com/rkt/stage1/gc" value:"/ex/gc" > > networks:<name:"default" ipv4:"172.16.28.2" > manifest:"{\"acVersion\":\"1.8.0+git98545e1\",\"acKind\":\"PodManifest\",\"apps\":[{\"name\":\"rkt-inspect\",\"image\":{\"name\":\"coreos.com/rkt-inspect\",\"id\":\"sha512-9ce35088a75eead90c15bd94736e0d3628c27a844398d0e3e9ff25e16242c878\",\"labels\":[{\"name\":\"version\",\"value\":\"1.8.0\"},{\"name\":\"arch\",\"value\":\"amd64\"},{\"name\":\"os\",\"value\":\"linux\"}]},\"app\":{\"exec\":[\"/inspect\",\"--read-stdin\"],\"user\":\"0\",\"group\":\"0\",\"workingDirectory\":\"/\",\"environment\":[{\"name\":\"VAR_FROM_MANIFEST\",\"value\":\"manifest\"},{\"name\":\"PATH\",\"value\":\"/bin\"}],\"isolators\":[{\"name\":\"resource/memory\",\"value\":{\"request\":\"25Mi\",\"limit\":\"25Mi\"}},{\"name\":\"resource/cpu\",\"value\":{\"request\":\"50m\",\"limit\":\"800m\"}}]},\"annotations\":[{\"name\":\"appc.io/docker/repository\",\"value\":\"whatever\"},{\"name\":\"coreos.com/rkt/stage1/run\",\"value\":\"/ex/run\"},{\"name\":\"coreos.com/rkt/stage1/enter\",\"value\":\"/ex/enter\"},{\"name\":\"coreos.com/rkt/stage1/gc\",\"value\":\"/ex/gc\"}]}],\"volumes\":null,\"isolators\":null,\"annotations\":null,\"ports\":[]}" cgroup:"/machine.slice/machine-rkt\\x2de363eae5\\x2da67f\\x2d4e4c\\x2dae8a\\x2d3f6b0f1aea0a.scope" created_at:1465568746015439946 started_at:1465568746143440523 
14:25:46 	rkt_api_service_test.go:159: Pod info: &{id:e363eae5-a67f-4e4c-ae8a-3f6b0f1aea0a pid:10515 state:running apps:map[rkt-inspect:0xc8203228e0] networks:map[default:0xc820302d00] manifest:0xc820276f00 createdAt:1465568746015000000 startedAt:1465568746143000000}
14:25:46 	rkt_api_service_test.go:611: Unexpected result: [], should see non-zero pods
14:25:46 	rkt_api_service_test.go:627: Unexpected result: [], should see non-zero pods
14:25:46 	rkt_api_service_test.go:156: API Pod info: id:"e363eae5-a67f-4e4c-ae8a-3f6b0f1aea0a" pid:10515 state:POD_STATE_EXITED apps:<name:"rkt-inspect" image:<base_format:<type:IMAGE_TYPE_APPC version:"0.8.4" > id:"sha512-9ce35088a75eead90c15bd94736e0d3628c27a844398d0e3e9ff25e16242c878" > annotations:<Key:"appc.io/docker/repository" value:"whatever" > annotations:<Key:"coreos.com/rkt/stage1/run" value:"/ex/run" > annotations:<Key:"coreos.com/rkt/stage1/enter" value:"/ex/enter" > annotations:<Key:"coreos.com/rkt/stage1/gc" value:"/ex/gc" > > created_at:1465568746015439946 started_at:1465568746143440523 
14:25:46 	rkt_api_service_test.go:159: Pod info: &{id:e363eae5-a67f-4e4c-ae8a-3f6b0f1aea0a pid:10515 state:exited apps:map[rkt-inspect:0xc820323340] networks:map[] manifest:0xc820277200 createdAt:1465568746015000000 startedAt:1465568746143000000}
```

The cgroup suffix of the Pod returned by the API service is the session one (`/user.slice/user-1000.slice/session-3.scope`) instead of having a `machine` scope (`machine-rkt\\x2de363eae5\\x2da67f\\x2d4e4c\\x2dae8a\\x2d3f6b0f1aea0a.scope`).

--- a/rkt/api_service.go
+++ b/rkt/api_service.go
@@ -16,6 +16,7 @@ package main
 
 import (
 	"encoding/json"
+	"errors"
 	"fmt"
 	"io/ioutil"
 	"net"
@@ -36,6 +37,7 @@ import (
 	"github.com/coreos/rkt/pkg/set"
 	"github.com/coreos/rkt/store"
 	"github.com/coreos/rkt/version"
+	"github.com/godbus/dbus"
 	"github.com/spf13/cobra"
 	"golang.org/x/net/context"
 	"google.golang.org/grpc"
@@ -368,6 +370,11 @@ func getBasicPod(p *pod) (*v1alpha.Pod, *schema.PodManifest) {
 	}
 
 	if pod.State == v1alpha.PodState_POD_STATE_RUNNING {
+		if err := waitForMachinedRegistration(pod.Id); err != nil {
+			// If there's an error, it means we're not registered to machined
+			// in a reasonable time. Just output the cgroup we're in.
+			stderr.PrintE("checking for machined registration failed", err)
+		}
 		// Get cgroup for the "name=systemd" controller.
 		pid, err := p.getContainerPID1()
 		if err != nil {
@@ -388,6 +395,25 @@ func getBasicPod(p *pod) (*v1alpha.Pod, *schema.PodManifest) {
 	}
 
 	return pod, manifest
+}
+
+func waitForMachinedRegistration(uuid string) error {
+	conn, err := dbus.SystemBus()
+	if err != nil {
+		return err
+	}
+	machined := conn.Object("org.freedesktop.machine1", "/org/freedesktop/machine1")
+	machineName := "rkt-" + uuid
+
+	var o dbus.ObjectPath
+	for i := 0; i < 10; i++ {
+		if err := machined.Call("org.freedesktop.machine1.Manager.GetMachine", 0, machineName).Store(&o); err == nil {
+			return nil
+		}
+		time.Sleep(time.Millisecond * 50)
+	}
+
+	return errors.New("pod not found")
 }
 
 func (s *v1AlphaAPIServer) ListPods(ctx context.Context, request *v1alpha.ListPodsRequest) (*v1alpha.ListPodsResponse, error) {

--- a/tests/rkt_api_service_test.go
+++ b/tests/rkt_api_service_test.go
@@ -21,8 +21,10 @@ import (
 	"encoding/json"
 	"fmt"
 	"os"
+	"os/user"
 	"path/filepath"
 	"reflect"
+	"strconv"
 	"strings"
 	"syscall"
 	"testing"
@@ -38,24 +40,34 @@ import (
 )
 
 func startAPIService(t *testing.T, ctx *testutils.RktRunCtx) *gexpect.ExpectSubprocess {
-	noGid := false
+	noUidGid := false
 	gid, err := common.LookupGid(common.RktGroup)
 	if err != nil {
 		t.Logf("no %q group, will run api service with root, ONLY DO THIS FOR TESTING!", common.RktGroup)
-		noGid = true
+		noUidGid = true
 	} else {
 		if err := ctx.SetupDataDir(); err != nil {
 			t.Fatalf("failed to setup data directory: %v", err)
 		}
 	}
 
+	u, err := user.Lookup("nobody")
+	if err != nil {
+		t.Logf(`no "nobody" user, will run api service with root, ONLY DO THIS FOR TESTING!`)
+		noUidGid = true
+	}
+	uid, err := strconv.Atoi(u.Uid)
+	if err != nil {
+		t.Fatalf(`failed to parse "nobody" UID: %v`, err)
+	}
+
 	t.Logf("Running rkt api service")
 	apisvcCmd := fmt.Sprintf("%s api-service", ctx.Cmd())
 
-	if noGid {
+	if noUidGid {
 		return startRktAndCheckOutput(t, apisvcCmd, "API service running")
 	}
-	return startRktAsGidAndCheckOutput(t, apisvcCmd, "API service running", gid)
+	return startRktAsUidGidAndCheckOutput(t, apisvcCmd, "API service running", false, uid, gid)
 }
 
 func stopAPIService(t *testing.T, svc *gexpect.ExpectSubprocess) {

--- a/tests/rkt_tests.go
+++ b/tests/rkt_tests.go
@@ -369,6 +369,29 @@ func startRktAsGidAndCheckOutput(t *testing.T, rktCmd, expectedLine string, gid 
 	return child
 }
 
+func startRktAsUidGidAndCheckOutput(t *testing.T, rktCmd, expectedLine string, expectError bool, uid, gid int) *gexpect.ExpectSubprocess {
+	child, err := gexpect.Command(rktCmd)
+	if err != nil {
+		t.Fatalf("cannot exec rkt: %v", err)
+	}
+	if gid != 0 {
+		child.Cmd.SysProcAttr = &syscall.SysProcAttr{}
+		child.Cmd.SysProcAttr.Credential = &syscall.Credential{Uid: uint32(uid), Gid: uint32(gid)}
+	}
+
+	err = child.Start()
+	if err != nil {
+		t.Fatalf("cannot start rkt: %v", err)
+	}
+
+	if expectedLine != "" {
+		if err := expectWithOutput(child, expectedLine); err != nil {
+			t.Fatalf("didn't receive expected output %q: %v", expectedLine, err)
+		}
+	}
+	return child
+}
+
 func runRktAndCheckRegexOutput(t *testing.T, rktCmd, match string) {
 	child := spawnOrFail(t, rktCmd)
 	defer child.Wait()

